### PR TITLE
Add initial rendering and tools to Model Rewrite

### PR DESCRIPTION
**Pull request by LeonMontealegre**

### CODE REVIEW

File: src/app/digital/internal/DigitalComponents.ts
Comment (LeonMontealegre) [inline src/app/digital/internal/DigitalComponents.ts:103]: Maybe consider a side "propagation-state" that's saved adjacent to the Schema as a "local state" or whatever

File: src/app/core/internal/view/CircuitView.ts
Comment (TGCrystal) [inline src/app/core/internal/view/CircuitView.ts:66]: TODOs like this should include a note about when this should be done. It is a requirement for this pr? A requirement for merging into master? Something that can wait for beyond that?

If you want to get real fancy with it, we can use this[ unicorn linting rule](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/expiring-todo-comments.md) but I don't think it is strictly necessary

(same goes for other TODOs on this file)

File: src/app/core/internal/view/rendering/RenderHelper.ts
Comment (TGCrystal) [inline src/app/core/internal/view/rendering/RenderHelper.ts:72]: Is this commented out code something to be TODOne later or should it be deleted?

File: src/app/core/internal/view/WireAssembler.ts
Comment (TGCrystal) [inline src/app/core/internal/view/WireAssembler.ts:13]: Should this somehow be assigned or connected with the `DefaultRenderOptions` below?

File: src/app/core/internal/view/rendering/prims/ArcCircle.ts
Comment (TGCrystal) [inline src/app/core/internal/view/rendering/prims/ArcCircle.ts:16]: I know this is probably copied and pasted over from master, but this `@param` description should be a little more descriptive or just nonexistent. I assume is it the position of where the center of the circle would be if it a full circle was drawn, but it could be misinterpreted as the position at which to start drawing from.

File: src/app/core/internal/view/rendering/prims/Curve.ts
Comment (TGCrystal) [inline src/app/core/internal/view/rendering/prims/Curve.ts:1]: These jsdoc comments could use a little bit of work. Just like ArcCircle, I assume this is largely copied over from master but there should be some elaboration on what "Control point" means or a link to something that explains it

File: src/app/core/internal/view/rendering/prims/SVG.ts
Comment (TGCrystal) [inline src/app/core/internal/view/rendering/prims/SVG.ts:43]: Is the `as string` cast here necessary? Can the condition check if `style.fill` is a string instead?

File: src/app/core/public/api/impl/Component.ts
Comment (TGCrystal) [inline src/app/core/public/api/impl/Component.ts:71]: Like the other TODOs, it may be helpful to elaborate when this should be done, but since this one is specifically directed towards you then it is your call

File: src/app/digital/internal/sim/Signal.ts
Comment (TGCrystal) [inline src/app/digital/internal/sim/Signal.ts:6]: Not sure if I'm misunderstanding the terminology, but wouldn't this actually be the `MetaUnstable` case?

File: src/app/digital/internal/views/DigitalCircuitView.ts
Comment (TGCrystal) [inline src/app/digital/internal/views/DigitalCircuitView.ts:31]: If this is going to include every digital component, I think the section should be grouped and ordered based on how they are in the itemnav (with DigitalWire at the beginning or end)

File: src/app/digital/internal/views/DigitalCircuitView.ts
Comment (TGCrystal) [inline src/app/digital/internal/views/DigitalCircuitView.ts:34]: Is there some way to use a union type rather than having to include this check? Or is that the issue we ran into with the first iteration of the model refactor?

File: src/app/digital/internal/views/DigitalWireAssembler.ts
Comment (TGCrystal) [inline src/app/digital/internal/views/DigitalWireAssembler.ts:28]: Think I mentioned this elsewhere on the pr, but shouldn't `"#ffffff"` be stored as `this.options.defaultOffColor`

File: src/app/digital/internal/views/components/ANDGateAssembler.ts
Comment (TGCrystal) [inline src/app/digital/internal/views/components/ANDGateAssembler.ts:32]: Any shot of type safety on this to avoid the `!`?

File: src/app/digital/internal/views/components/SwitchAssembler.ts
Comment (TGCrystal) [inline src/app/digital/internal/views/components/SwitchAssembler.ts:23]: Magic numbers, I don't want to have to figure out what these mean if/when I implement that eslint rule at some point in the future

File: src/app/digital/internal/views/components/SwitchAssembler.ts
Comment (TGCrystal) [inline src/app/digital/internal/views/components/SwitchAssembler.ts:41]: Another magic number

File: src/site/pages/digital/src/index.tsx
Comment (TGCrystal) [inline src/site/pages/digital/src/index.tsx:189]: Add a TODO on when these will be uncommented

File: src/site/shared/containers/MainDesigner/index.tsx
Comment (TGCrystal) [inline src/site/shared/containers/MainDesigner/index.tsx:29]: Add TODO to uncomment/reenable

File: src/site/shared/utils/GenerateThumbnail.ts
Comment (TGCrystal) [inline src/site/shared/utils/GenerateThumbnail.ts:15]: Add TODO to uncomment/reenable

**Review by TGCrystal — CHANGES_REQUESTED** [review]
Don't think it is strictly necessary, but I think anything in a public folder should have jsdoc comments, and it is better you writing them than someone in the future who only partially understands how it all works. If you want I can set up an eslint rule for this

File: src/app/core/internal/view/CircuitView.ts
Comment (LeonMontealegre) [reply src/app/core/internal/view/CircuitView.ts:66]: Yeah... you're definitely out-ing me here, I'll try and come up with a standard

File: src/app/core/internal/view/WireAssembler.ts
Comment (LeonMontealegre) [reply src/app/core/internal/view/WireAssembler.ts:13]: Yeah, it uses `this.options` which is an instance of `DefaultRenderOptions`

File: src/app/core/internal/view/rendering/prims/ArcCircle.ts
Comment (LeonMontealegre) [reply src/app/core/internal/view/rendering/prims/ArcCircle.ts:16]: 👺

File: src/app/core/internal/view/rendering/prims/SVG.ts
Comment (LeonMontealegre) [reply src/app/core/internal/view/rendering/prims/SVG.ts:43]: It's necessary because it could be a `CanvasGradient`, but yeah i should probably throw an error in that case

File: src/app/digital/internal/sim/Signal.ts
Comment (LeonMontealegre) [reply src/app/digital/internal/sim/Signal.ts:6]: It's just called Metastability, that's just the term I believe

File: src/app/digital/internal/views/components/SwitchAssembler.ts
Comment (LeonMontealegre) [reply src/app/digital/internal/views/components/SwitchAssembler.ts:23]: The magic numbers will have to exist somewhere, this is where - I'm trying to stray away from a monolithic `Constants.ts` file

I can make them static I guess

Comment (TGCrystal) [reply src/app/digital/internal/views/components/SwitchAssembler.ts:23]: What about having the constants defined in this file instead of a monolithic Constants.ts file? Just something like `const SWITCH_HORIZONTAL_SIZE = 1.24` or something like that? Or is that what you mean by making them static?

Comment (LeonMontealegre) [reply src/app/digital/internal/views/components/SwitchAssembler.ts:23]: An example of bad magic numbers is:
```ts
        if (transformChanged) {
            // Update transform
            this.view.componentTransforms.set(sw.id, new Transform(
                V(sw.props.x ?? 0, sw.props.y ?? 0),
                V(1.24, 1.54), <-------------
                (sw.props.angle ?? 0),
            ));
        }
```
These are named and their meaning is plenty obvious, putting them in a constant doesn't add anything here

File: src/app/core/internal/view/CircuitView.ts
Comment (LeonMontealegre) [reply src/app/core/internal/view/CircuitView.ts]: Okay, standard is kinda-set:
`TODO[the-branch-to-deal-with-this-in](the-person-in-charge)`, i.e.
`TODO[model_refactor](leon)` means this TODO needs to be dealt with before `model_refactor` can be merged, and Leon is in charge of it

**Review by KevinMackenzie — APPROVED** [review]
Looks fine for a v0.1, no major issues.  Approving to unblock other work.

**Review by TGCrystal — APPROVED** [review]

**Pull request merged**
